### PR TITLE
fixes #13131: support matrix parameters and other parameter styles in typescript-angular

### LINKED ISSUE #13131: [REQ] support matrix parameters and other parameter styles in typescript-angular
### Is your feature request related to a problem? Please describe.

I need to access a server that is making heavy use of matrix parameters (i.e.: "in": "path", "style": "matrix").
At least the "typescript-angular" generator does not support any other parameter-styles than simple.

The way parameter encoding currently is implemented, modifying the mustache-templates does not work:
AbstractTypeScriptClientCodegen.java generates some hardcoded Typescript code that is broken for other styles than "simple".

Other typescript-* generators are probably "broken" in the same way since they make use of the same Java abstract generator class, see above.

## Describe the solution you'd like

I'd like to have either
a) have correct code generated for all supported parameter styles (at least "style": "matrix")  or
b) have code generated in a way that allows for customization via mustache-templates.

## Describe alternatives you've considered

I tried to customize api.service.mustache but the relevant/wrong code is hardcoded in AbstractTypeScriptClientCodegen.java.

## Additional context

n/a/

### PR DESCRIPTION
**Pull request by HonoluluHenk**
Optionally generate code that allows for customization of parameter encoding in angular-typescript.

If this gets accepted, I'm willing to integrate this into other typescript generators.

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (6.1.0) (minor release - breaking changes with fallbacks), `7.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

Mentions:
@TiFu  @taxpon @sebastianhaas @kenisteward @Vrolijkx  @macjohnny @topce @akehir @petejohansonxo @amakhrov  @davidgamero @mkusaka

### CODE REVIEW

**Comment by HonoluluHenk** [comment]
Running static code analysys errors out on master, so I did not run this.

**Comment by HonoluluHenk** [comment]
fixes #13131

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java [hunk 1]
```diff
@@ -39,13 +39,117 @@
...
 public abstract class AbstractTypeScriptClientCodegen extends DefaultCodegen implements CodegenConfig {
+
+    @SuppressWarnings("StringBufferField")
+    public static class ParameterExpander {
+        public enum STRATEGY {
+            legacy("Old legacy implementation that only supports the 'simple' style"
+                           + " and only calls encodeURIComponent()"
+                           + " (as opposed to encoding of all restricted characters of RFC 6570)"),
+            custom("Allows for providing a custom parameter encoding function via customization of " +
+                           "the 'api.service.mustache' template.");
```
Comment (macjohnny) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java:58]: I suggest to create a separate .mustache template, so that a consumer does not need to copy the entire api.service.mustache in order to customize that.

As an alternative, can we use the `Configuration` object to pass suche a custom parameter encoding function?

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java [hunk 2]
```diff
@@ -39,13 +39,117 @@
...
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 import static org.openapitools.codegen.utils.StringUtils.camelize;
 import static org.openapitools.codegen.utils.StringUtils.underscore;
 
 public abstract class AbstractTypeScriptClientCodegen extends DefaultCodegen implements CodegenConfig {
+
+    @SuppressWarnings("StringBufferField")
+    public static class ParameterExpander {
```
Comment (macjohnny) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java:52]: can you add some usage example to the Readme.mustache?

**Comment by macjohnny** [comment]
@HonoluluHenk thanks for your contribution!

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java (hunk 1 above)
Comment (HonoluluHenk) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java:58]: > can we use the Configuration object to pass such a custom parameter encoding function?

veeeery nice idea, didn't think of that :facepalm: 

Lets say, I'd create the Configuration property `paramEncoder?: (param: Parameter) => string;`

There are multiple ways of implementing this: 
1. *keep* the new Cli-Param `paramExpansionStrategy` and only generate the encoder-stuff if the cli-param is set.
2. *drop* the new Cli-Param and evaluate the Configuration property mentioned above
  a) implement this in a way that api.service.mustache is still unchanged (this will make the java code uglier)
  b) api.service.mustache may be changed so that the new behavior is implemented nicely

Option 1 might be implemented in a way so current generator setups may run unmodified. As such, this might be released in the next minor version.
Option 2 is (imho) a breaking change as customized mustache templates need to be rebased to the new versions (at the very least configuration.mustache and probably api.service.mustache)

Depending how much you value 100% backwards compatibility vs future usability vs maintainability, all 3 options provide benefits and drawbacks.

Please choose :)

P.S.: I'd be happy to implement multiple versions if you want to try different approaches!

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java (hunk 2 above)
Comment (HonoluluHenk) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java:52]: done, see commit 6eefcbd

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java (hunk 1 above)
Comment (macjohnny) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java:58]: thanks for evaluating the options

I would go with option 2b, since overriding templates is not guaranteed to be backwards compatible

**Comment by HonoluluHenk** [comment]
awaiting next review...

File: modules/openapi-generator/src/main/resources/typescript-angular/README.mustache [hunk 3]
```diff
@@ -201,3 +201,21 @@ import { environment } from '../environments/environment';
 })
 export class AppModule { }
 ```
+
+### Customizing parameter encoding
+
+Without further customization, only parameters of style 'simple' are encoded correctly.
+
+Other styles ("matrix", "label", "form", ...) are not that easy to encode and thus are best delegated to
```
Comment (macjohnny) [inline modules/openapi-generator/src/main/resources/typescript-angular/README.mustache:209]: can you link the openapi spec, e.g. https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#parameter-locations?

File: modules/openapi-generator/src/main/resources/typescript-angular/configuration.mustache [hunk 4]
```diff
@@ -14,6 +15,7 @@ export interface {{configurationParametersInterfaceName}} {
     basePath?: string;
     withCredentials?: boolean;
     encoder?: HttpParameterCodec;
+    encodeParam?: (param: Param) => string;
```
Comment (macjohnny) [inline modules/openapi-generator/src/main/resources/typescript-angular/configuration.mustache:18]: can you please add docs?

File: modules/openapi-generator/src/main/resources/typescript-angular/configuration.mustache [hunk 5]
```diff
@@ -157,4 +168,16 @@ export class {{configurationClassName}} {
             ? value()
             : value;
     }
+
+    private legacyEncodeParam(param: Param): string {
+        // This legacy implementation is here for backwards compatibility only.
+        // It is only correct for param-style 'simple'.
+        if(param.style === "simple") {
```
Comment (macjohnny) [inline modules/openapi-generator/src/main/resources/typescript-angular/configuration.mustache:175]:
```suggestion
        if (param.style === "simple") {
```

File: modules/openapi-generator/src/main/resources/typescript-angular/configuration.mustache [hunk 6]
```diff
@@ -157,4 +168,16 @@ export class {{configurationClassName}} {
             ? value()
             : value;
     }
+
+    private legacyEncodeParam(param: Param): string {
+        // This legacy implementation is here for backwards compatibility only.
```
Comment (macjohnny) [inline modules/openapi-generator/src/main/resources/typescript-angular/configuration.mustache:173]: isn't this also just the default encoding function if none is given?

File: modules/openapi-generator/src/main/resources/typescript-angular/configuration.mustache [hunk 7]
```diff
@@ -157,4 +168,16 @@ export class {{configurationClassName}} {
...
+        // It is only correct for param-style 'simple'.
+        if(param.style === "simple") {
+            return encodeURIComponent(String(param.value));
+        }
```
Comment (macjohnny) [inline modules/openapi-generator/src/main/resources/typescript-angular/configuration.mustache:177]: please also add the date handling from 

https://github.com/OpenAPITools/openapi-generator/blob/2a196ce3620aebf9e5ce0c9a17ee8e0c5365afb7/modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/TypeScriptAngularClientCodegen.java#L428

File: modules/openapi-generator/src/main/resources/typescript-angular/param.mustache [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+  | 'simple'
+  | 'spaceDelimited'
+  | 'pipeDelimited'
+  | 'deepObject'
+  ;
+
+/**
+ * Convenience: simulate ParameterStyle enum
+ */
+export const ParamStyle = {
```
Comment (macjohnny) [inline modules/openapi-generator/src/main/resources/typescript-angular/param.mustache:17]: nice!

how about exporting the type as a string enum

```typescript
enum ParamStyle {
  matrix,
  ...
}
```
?

see the discussion in https://github.com/OpenAPITools/openapi-generator/issues/3451

Comment (macjohnny) [reply modules/openapi-generator/src/main/resources/typescript-angular/param.mustache:17]: same below

File: modules/openapi-generator/src/test/java/org/openapitools/codegen/testutils/AssertFile.java [hunk 9]
```diff
@@ -84,9 +84,10 @@ public FileVisitResult preVisitDirectory(Path expectedDir, BasicFileAttributes a
                         Arrays.sort(actual);
                     }
 
-                    assertEquals(expected,
-                            actual,
-                            String.format(Locale.ROOT, "Directory content of '%s' and '%s' differ.", expectedDir, actualDir));
+                    // Use AssertJ as it also prints a diff
+                    assertThat(actual)
+                            .describedAs(String.format(Locale.ROOT, "Directory content of '%s' and '%s' differ.", expectedDir, actualDir))
+                            .containsExactly(expected);
```
Comment (macjohnny) [inline modules/openapi-generator/src/test/java/org/openapitools/codegen/testutils/AssertFile.java:90]: is this change related to this PR?

File: modules/openapi-generator/src/test/java/org/openapitools/codegen/typescript/TypeScriptGroups.java [hunk 10]
```diff
@@ -0,0 +1,12 @@
+package org.openapitools.codegen.typescript;
+
+public final class TypeScriptGroups {
```
Comment (macjohnny) [inline modules/openapi-generator/src/test/java/org/openapitools/codegen/typescript/TypeScriptGroups.java:3]: please add docs

File: modules/openapi-generator/src/test/java/org/openapitools/codegen/typescript/typescriptangular/TypescriptAngularParamsStrategyCustomIntegrationTest.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2018 OpenAPI-Generator Contributors (https://openapi-generator.tech)
+ * Copyright 2018 SmartBear Software
```
Comment (macjohnny) [inline modules/openapi-generator/src/test/java/org/openapitools/codegen/typescript/typescriptangular/TypescriptAngularParamsStrategyCustomIntegrationTest.java:3]:
```suggestion
```

**Review by macjohnny — COMMENTED** [review]
thanks for your contribution!

File: modules/openapi-generator/src/test/java/org/openapitools/codegen/testutils/AssertFile.java (hunk 9 above)
Comment (HonoluluHenk) [reply modules/openapi-generator/src/test/java/org/openapitools/codegen/testutils/AssertFile.java:90]: Yes and no...
If there are additional or missing files, the old implementation simply does not print /which/ file is missing/surplus but just that "files differ".
One evening, I spent a *lot* of time and nerves searching for a missing file :(

AssertJ gives this information out of the box (assertion behavior otherwise is the same as before)

Removing is no problem (you decide).
In case of removal: if you see just a slight chance that this change might get incorporated, I'd be glad to open another PR just for this because it is just soooo annoying :)

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java [hunk 12]
```diff
@@ -39,13 +39,110 @@
...
```
[the remark this answers is not in the document]
Comment (HonoluluHenk) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java:53]: Btw: is there a policy that Code-Generators may not depend on other classes?

I'd rather have this a standalone class in package (e.g.) org.openapitools.codegen or .languages but refrained since this would be the only class that is not a Code-Generator of one of the abstract base classes

File: modules/openapi-generator/src/main/resources/typescript-angular/param.mustache (hunk 8 above)
Comment (HonoluluHenk) [reply modules/openapi-generator/src/main/resources/typescript-angular/param.mustache:17]: Hmm... don't forget about the associated union-type:
```typescript
export type PathParameterStyle =
  | 'matrix'
  | 'label'
...
```

This allows me to write easy to read code:
```typescript
const style: ParamStyle = 'matrix';
```

Using enums, users'd have to write the loooooong (and thus annoying):
```typescript
const style: ParamStyle = ParamStyle.matrix;
```


The const object is solely there if anyone wants to iterate over all possible `ParamStyle`s and could be dropped without loss.
Please also note: it's exported `as const` so it's effectively the same as the enum.
And finally: enums are nowhere to be found in @angular/core. (union-types all the way, iirc).

If I'd have to choose, I'd rather keep the union and drop the const (without introducing the enum).

But: your decision, please advise.

Comment (HonoluluHenk) [reply modules/openapi-generator/src/main/resources/typescript-angular/param.mustache:17]: p.s.: afaik, you cannot have an enum + union type: this does not compile because enum types cannot be merged with other types, just other enums

**Comment by HonoluluHenk** [comment]
@macjohnny thanks for your thorough reviews!

Awaiting the next review and discussions...

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (macjohnny) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java:73]: I am not aware of a policy that would disallow depending on other classes. 

we can keep it in the abstract class, but if you want, we could create a sub package typescriptutils or something like that, and move it to a standalone class there. Up to you (you could do that in a separate PR)

File: modules/openapi-generator/src/main/resources/typescript-angular/param.mustache (hunk 8 above)
Comment (macjohnny) [reply modules/openapi-generator/src/main/resources/typescript-angular/param.mustache:22]: Since string enums should become the default for the genrated angular code, but currently its still not the default:

https://github.com/OpenAPITools/openapi-generator/blob/18bb5466bb3c32f873337a8bc9f2fe6687490cb4/modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/TypeScriptAngularClientCodegen.java#L80

I suggest to use string enums here.

> enums are nowhere to be found in @angular/core. (union-types all the way, iirc).

I wonder whether this has historical reasons (pre TS 2.4) or whether there is another compelling reason. If you find something, let me know, in the end either way would work for the code here.

File: modules/openapi-generator/src/test/java/org/openapitools/codegen/testutils/AssertFile.java (hunk 9 above)
Comment (macjohnny) [reply modules/openapi-generator/src/test/java/org/openapitools/codegen/testutils/AssertFile.java:90]: Since I have already reviewed it, lets keep it in this PR.

For future PRs, I am happy to review them, and I am even happier if they are small, so I can look at them whenever I have some spare minutes ;)

**Comment by macjohnny** [comment]
@HonoluluHenk anything else you need from me before I should have a final look at it?
Btw thanks for your patience and contribution, much appreciated. Sorry for the review delays, sometimes its hard to find time :)

**Comment by HonoluluHenk** [comment]
@macjohnny 
> Sorry for the review delays

Lol... a PR of mine for another project just got merged after 3 years of waiting... I'm fine ;).
Actually, only few high-profile projects are this responsive, I'm quite impressed!

After doing some research and POCs, I removed the enums completely, because:
* unexpectedly, It would require a lot of boilerplate code to parse predefined values into Typescript-enum values (example: the format 'date-time' is not a valid enum key-identifier and thus would require some parsing/mapping logic)... for absolutely no gain in Java-maintainability or Typescript usability/type-safety.
* it _just does not feel Angular-ish_ (see comments above: Angular is union-types-all-the-way)

If there are no more issues on your side, that's all for this PR/issue from me.

Again, thanks for your professional reviews and patience with the back and forth!

**Review by macjohnny — APPROVED** [review]
LGTM
Thanks!

**Comment by macjohnny** [comment]
@HonoluluHenk can you fix the failing CI tests? Then the PR is ready to merge

**Comment by HonoluluHenk** [comment]
@macjohnny uhm... I actually have no Idea which branch/commit the CI is building:

The Unit tests build log says it is [checking out commit c79cad4a17a0d2e87512b2817b35e2bfa3987a1a](https://github.com/OpenAPITools/openapi-generator/runs/7969302925?check_suite_focus=true#step:2:526)

I cannot find this commit nowhere in upstream, my fork or my local repo.
The last commit in my repo/fork (and documented here in this PR, see above) is: 24d725af (... this commit addressed all the failures bleated out by CI)

Any help appreciated on where to search for the missing commit or how to mitigate.

**Comment by macjohnny** [comment]
@HonoluluHenk can you merge the latest master into your branch and re-generate the samples? This usually resolves the issue

**Comment by HonoluluHenk** [comment]
Thanks, that did the trick.
All green now...
